File: 2018/knowit/6/6.hs
isZeroHeavy :: Int -> Bool
isZeroHeavy n = isZeroHeavy' (0, n)
    where isZeroHeavy' (balance, 0) = balance > 0
          isZeroHeavy' (balance, n)
              | n `mod` 10 == 0 = isZeroHeavy' (balance+1, m)
              | otherwise       = isZeroHeavy' (balance-1, m)
              where m = n `div` 10

main = putStrLn
    . show
    . sum
    . filter isZeroHeavy
    $ [1..18163106]

File: 2018/knowit/6/6.js
"use strict";

console.time("Solved");

function isZeroHeavy(n) {
	return [...n.toString()].reduce((balance, c) => c === '0' ? balance+1 : balance-1, 0) > 0;
}

let sum = 0;

for (let i = 0; i <= 18163106; i++) {
	if (isZeroHeavy(i)) {
		sum += i;
	}
}

console.timeEnd("Solved");

console.log(sum);
